app/utils/IncomeSourcesUtils.scala

```diff
@@ -0,0 +1,128 @@
+import play.api.mvc.Session
+
+import java.time.LocalDate
+
+case class MissingKey(msg: String)
+
+object IncomeSourcesUtils {
+
+  def getBusinessDetailsFromSession(implicit user: MtdItUser[_]): Either[Throwable, CheckBusinessDetailsViewModel] = {
+    val userActiveBusinesses: List[BusinessDetailsModel] = user.incomeSources.businesses.filterNot(_.isCeased)
+    val skipAccountingMethod: Boolean = userActiveBusinesses.isEmpty
+    val errors: Seq[String] = Seq(
+      user.session.data.get(businessName).orElse(Option(MissingKey("MissingKey: addBusinessName"))),
+      user.session.data.get(businessStartDate).orElse(Option(MissingKey("MissingKey: addBusinessStartDate"))),
+      user.session.data.get(businessTrade).orElse(Option(MissingKey("MissingKey: addBusinessTrade"))),
+      user.session.data.get(addBusinessAddressLine1).orElse(Option(MissingKey("MissingKey: addBusinessAddressLine1"))),
+      user.session.data.get(addBusinessPostalCode).orElse(Option(MissingKey("MissingKey: addBusinessPostalCode")))
+    ).collect {
+      case Some(MissingKey(msg)) => MissingKey(msg)
+    }.map(e => e.msg)
+
+
+    val result: Option[CheckBusinessDetailsViewModel] = for {
+      businessName <- user.session.data.get(businessName)
+      businessStartDate <- user.session.data.get(businessStartDate).map(LocalDate.parse)
+      businessTrade <- user.session.data.get(businessTrade)
+      businessAddressLine1 <- user.session.data.get(addBusinessAddressLine1)
+      businessAccountingMethod <- user.session.data.get(addBusinessAccountingMethod)
+      accountingPeriodEndDate <- user.session.data.get(addBusinessAccountingPeriodEndDate).map(LocalDate.parse)
+    } yield {
+
+      CheckBusinessDetailsViewModel(
+        businessName = Some(businessName),
+        businessStartDate = Some(businessStartDate),
+        accountingPeriodEndDate = accountingPeriodEndDate,
+        businessTrade = businessTrade,
+        businessAddressLine1 = businessAddressLine1,
+        businessAddressLine2 = user.session.data.get(addBusinessAddressLine2),
+        businessAddressLine3 = user.session.data.get(addBusinessAddressLine3),
+        businessAddressLine4 = user.session.data.get(addBusinessAddressLine4),
+        businessPostalCode = user.session.data.get(addBusinessPostalCode),
+        businessCountryCode = user.session.data.get(addBusinessCountryCode),
+        businessAccountingMethod = user.session.data.get(addBusinessAccountingMethod),
+        cashOrAccrualsFlag = businessAccountingMethod,
+        skippedAccountingMethod = skipAccountingMethod)
+    }
+
+    result match {
+      case Some(checkBusinessDetailsViewModel) =>
+        Right(checkBusinessDetailsViewModel)
+      case None =>
+        Left(new IllegalArgumentException(s"Missing required session data: ${errors.mkString(" ")}"))
+    }
+  }
+
+  def getUKPropertyDetailsFromSession(implicit user: MtdItUser[_]): Either[Throwable, CheckUKPropertyViewModel] = {
+    val errors: Seq[String] = Seq(
```

Comment:
this processing is done unnecessarily in the happy path. Perhaps we could move it to the result -> None case block?

Reply:
lazy val might achieve the same tbf